Implement book cover feature

damgam0288 opened the pull request:
### Description

This PR is a first iteration of the feature request for a book cover in JabRef in [issue 10120](https://github.com/JabRef/jabref/issues/10120).

### Changes

When users add an image as a File for a Book, In-book or Booklet citation with the description "cover", it will appear on the right side of the Preview panel.

<img width="547" alt="PR-10120-edit-file-dialog-box" src="https://github.com/user-attachments/assets/7f19a9ba-113b-4d0d-9719-17965b42500e">

<img width="1128" alt="PR-10120-show-book-cover" src="https://github.com/user-attachments/assets/49c4b0a4-d2d5-4918-ad1d-138ead9185a8">

### Code changes

#### `BibEntry.java`

- `getCoverImageFile` method - searches the Entry's files for an image with the `COVER_TAG` in the description
- `COVER_TAG` string field - the text that users must include in the file's description to be detected as the book cover
- `isCoverable` method - checks if given Entry type can have a cover image `COVERABLE_TYPES`
- `COVERABLE_TYPES` field - a list of BibEntry types that can have a book cover (at the moment, this is `Book`, `In-book` and `Booklet`)

#### `PreviewViewer.java`

- `setPreviewText` method - edited the HTML to include an `img` tag for the book cover
- `getBookCoverURI` method - supplies the source for the `img` tag mentioned above

#### `BibEntryTest.java`

- `getCoverImageReturnsEmptyIfNoFiles`
- `getCoverImageReturnsEmptyIfNoImageFiles`
- `getCoverImageReturnsEmptyIfEntryIsNotCoverable`
- `getCoverImageDoesNotReturnImagesWithoutCoverDescription`
- `getCoverImageDoesNotReturnDocumentsWithCoverDescription`
- `getCoverImageReturnsCorrectImage`
- `getCoverImageUpdatesWithChangeToDescription`

### Notes

- I am not sure how to approach allowing the user to toggle the book cover on/off because it is part of the HTML in `setPreviewText` method.
- I have implemented this feature initially by relying on the user to select the image but I would like to use the [book cover API](https://github.com/w3slley/bookcover-api) as per the [issue discussion](https://github.com/JabRef/jabref/issues/10120#issuecomment-2419062484) and would appreciate some advice on how to do this without harming the code-base.

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- N/A [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- N/A [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

damgam0288 commented on src/main/java/org/jabref/model/entry/BibEntry.java:
```diff
@@ -55,6 +55,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import scala.annotation.meta.field;
```
Wrong import sorry, will remove

damgam0288 commented on src/main/java/org/jabref/model/entry/LinkedFile.java:
```diff
@@ -220,6 +223,10 @@ public boolean isOnlineLink() {
     }
 
+    public boolean isImage() {
+        return IMAGE_EXTENSIONS.stream().anyMatch(getFileType().toLowerCase()::contains);
```
I've discovered that exclusively image files have the `String` "image" in their `fileTypeProperty` - should I just check for that `String` instead of using `HashSet` altogether?

subhramit commented:
@damgam0288 How much is left to be done here? Once you are done, please fix the openrewrite tests and re-request a review for further approval.

damgam0288 commented:
> @damgam0288 Hey, how much is left to be done here? Once you are done, please fix the openrewrite tests and re-request a review for further approval.

No worries, I've finished the work best I can, and pushed to the branch. I'll fix the openrewrite tests in the next few hours.

damgam0288 commented:
@subhramit  Can you help please - is there anyway to run CheckRewrite without pushing to the branch? I am not sure what part of my code is causing it to fail - the report says "BibEntryTest" - I am trying to fix it but I think I am cluttering up the commit history unnecessarily.

subhramit commented:
> @subhramit  Can you help please - is there anyway to run `CheckRewrite` without pushing to the branch? I am not sure what part of my code is causing it to fail - the report says `BibEntryTest` - I am trying to fix it but I think I am cluttering up the commit history unnecessarily in my attempts to find the problem.

Run the command `gradlew rewriteRun`
Or double click the gradle task as per https://devdocs.jabref.org/code-howtos/faq.html#failing-openrewrite-tests

subhramit commented:
@damgam0288 right now, checkstyle test is red as well. You can take help from the same FAQ page about what to do. You have to have your IntelliJ configured for the local checkstyle warnings to see which part of the code is causing it.
For openrewrite, running the command/task is sufficient to fix it.

damgam0288 commented:
> @damgam0288 right now, checkstyle test is red as well. You can take help from the same FAQ page about what to do. You have to have your IntelliJ configured for the local checkstyle warnings to see which part of the code is causing it. For openrewrite, running the command/task is sufficient to fix it.

No worries, that checkstyle error was a result of a previous trial-and-error attempt with `OpenRewrite` - it's an unused import I will fix with my next commit. Thank you so much for the help with `OpenRewrite` :) - I was using `terminal` to run gradle so it wasn't working correctly. I found the errors and I'm working on them now.

subhramit commented:
Please revert your last commit, and use the gradle sidebar (double click on the task) to fix it. Your JDK is outdated, so using `rewriteRun` is affecting multiple classes.

subhramit commented:
I removed it from the drafts, hope that's fine?

damgam0288 commented:
> I removed it from the drafts, hope that's fine?

Of course, no problem 👍

Sorry I missed your last message but I think I fixed the rewrite issues? Or is it still a problem?

subhramit commented:
> Sorry I missed your last message but I think I fixed the rewrite issues? Or is it still a problem?

No, all good now. Thanks.

damgam0288 commented:
I am re-writing my tests to suit OpenRewrite, then will re-request review

subhramit commented:
> I am re-writing my tests to suit OpenRewrite, then will re-request review

You don't have to manually "suit" openrewrite when coding. As long as your code works, running the task will refactor it as per openrewrite. You can then just commit and push changes.

damgam0288 commented:
The following test was causing an `IndexOutOfBoundsException` from `OpenRewrite`. I tried my best but I couldn't fix it. It might be a bug because I found [this issue](https://github.com/openrewrite/rewrite/issues/3737) on the GitHub repository for OpenRewrite. Any suggestions on what to do? Thank-you.

  ```
 @Test
    void getCoverImageUpdatesWithChangeToDescription() {
        List<LinkedFile> files = new ArrayList<>();

        files.add(new LinkedFile("cover", Paths.get("JabRef-icon-128.png"), "PNG image"));
        files.add(new LinkedFile("", Paths.get("JabRef-icon-64.png"), "PNG image"));
        LinkedFile cover1 = files.get(0);
        LinkedFile cover2 = files.get(1);

        BibEntry entry = new BibEntry(StandardEntryType.Book).withField(StandardField.AUTHOR, "value");
        entry.setFiles(files);

        assertEquals(Optional.of(cover1), entry.getCoverImageFile());

        cover1.setDescription("");
        cover2.setDescription("cover");
        entry.setFiles(files);

        assertEquals(Optional.of(cover2), entry.getCoverImageFile());
    }
```

koppor reviewed (changes requested):
While reviewing, it came clear to me that the "architecture" of the solution is wrong.

`isImage` is good, the other things maybe not. We need to try out. In your solution, the types with a cover are hard-coded. I would go the extreme other way and show the first image as cover in all cases... Let me describe below:

JabRef offers configurability.  For the preview, this is https://docs.jabref.org/setup/preview Read about the syntax at https://docs.jabref.org/collaborative-work/export/customexports.

JabRef should also be "knobless". It should "magically" do something if possible - and not force the user.

Steps:

1. Understand `FileLink(filetype)`. Implementation is at org.jabref.logic.layout.format.FileLink#format
2. Add a jpg image to an entry
3. Adapt JabRef prview preferences to use filelink to show that image
4. Add a png image to another entry
5. Adapt JabReff preview preferences to use filelink+png to show that image
6. Adapt FileLink to support "image" as general file type -- it then searches for all image types and returns the first hit. This makes JabRef "knobless", because JabRef automatically finds the image
7. Adapt `        defaults.put(PREVIEW_STYLE,` in `org.jabref.gui.preferences.JabRefGuiPreferences#JabRefGuiPreferences` to contain your adaption
8. Adapt `org.jabref.migrations.PreferencesMigrations#upgradePreviewStyle` to migrate to your new style

I think, step 3 is the most challenging

---

Screenshot of the preferences:

![image](https://github.com/user-attachments/assets/4dee219f-3a9e-4fe5-b56c-d628c5fcf4f4)

koppor commented:
Closing this issue due to inactivity :zzz:
